plugin/db/util/driverutil.go

```diff
@@ -216,36 +203,40 @@ func beginMigration(ctx context.Context, executor MigrationExecutor, m *db.Migra
 		}
 	}
 
-	// Check if there is any higher version already been applied
-	if version, err := executor.CheckOutOfOrderVersion(ctx, tx, m.Namespace, m.Source, m.Version); err != nil {
+	sqldb, err := executor.GetDbConnection(ctx, bytebaseDatabase)
+	if err != nil {
+		return -1, err
+	}
+	tx, err := sqldb.BeginTx(ctx, nil)
+	if err != nil {
 		return -1, err
-	} else if version != nil && len(*version) > 0 {
-		// Clickhouse will always return non-nil version with empty string.
-		return -1, common.Errorf(common.MigrationOutOfOrder, fmt.Errorf("database %q has already applied version %s which is higher than %s", m.Database, *version, m.Version))
-	}
-
-	// If the migration engine is VCS and type is not baseline and is not branch, then we can only proceed if there is existing baseline
-	// This check is also wrapped in transaction to avoid edge case where two baselinings are running concurrently.
-	if m.Source == db.VCS && m.Type != db.Baseline && m.Type != db.Branch {
-		if hasBaseline, err := executor.FindBaseline(ctx, tx, m.Namespace); err != nil {
-			return -1, err
-		} else if !hasBaseline {
-			return -1, common.Errorf(common.MigrationBaselineMissing, fmt.Errorf("%s has not created migration baseline yet", m.Database))
-		}
 	}
+	defer tx.Rollback()
 
-	// VCS based SQL migration requires existing baselining
-	requireBaseline := m.Source == db.VCS && m.Type == db.Migrate
-	sequence, err := executor.FindNextSequence(ctx, tx, m.Namespace, requireBaseline)
+	latestSequence, err := executor.FindLatestSequence(ctx, tx, m.Namespace)
 	if err != nil {
 		return -1, err
 	}
 
-	// Phase 2 - Record migration history as PENDING
+	// Check if there is any higher version already been applied.
+	if latestSequence > 0 && m.Type != db.Baseline && m.Type != db.Branch {
+		if list, err := executor.FindMigrationHistoryList(ctx, &db.MigrationHistoryFind{
+			Database: &m.Namespace,
+			Sequence: &latestSequence,
+		}); err != nil {
+			return -1, fmt.Errorf("Find latest migration error: %q", err)
+		} else if len(list) > 0 {
+			if list[0].Version > m.Version {
+				return -1, common.Errorf(common.MigrationOutOfOrder, fmt.Errorf("database %q has already applied the latest version %s which is higher than %s", m.Database, list[0].Version, m.Version))
+			}
+		}
+	}
+
+	// Phase 2 - Record migration history as PENDING.
 	// MySQL runs DDL in its own transaction, so we can't commit migration history together with DDL in a single transaction.
 	// Thus we sort of doing a 2-phase commit, where we first write a PENDING migration record, and after migration completes, we then
 	// update the record to DONE together with the updated schema.
-	if insertedID, err = executor.InsertPendingHistory(ctx, tx, sequence, prevSchema, m, statement); err != nil {
+	if insertedID, err = executor.InsertPendingHistory(ctx, tx, latestSequence+1, prevSchema, m, statement); err != nil {
```

Comment:
I guess we no longer need the transaction since latestSequence+1 will fail UNIQUE constraint if there are 2 concurrent insertion.

Reply:
Yeah, but we can still use the same transaction for coding-wise simplicity here.

Reply:
Another problem to solve is to prevent two concurrent transactions triggered by Bytebase.

Reply:
>> Yeah, but we can still use the same transaction for coding-wise simplicity here.

I mean we can get rid of the beginTx/commit and do not need to expose the tx parameter in the interface.

>>  Another problem to solve is to prevent two concurrent transactions triggered by Bytebase.

It can, but the unique (namespace, sequence) constraint should protect us?

Reply:
I mean the task would still fail and cause the workflow to fail.

https://linear.app/bbteam/issue/BYT-328/integration-test-can-fail-with-database-xxhas-already-applied-version

Reply:
I still fail to understand why explicitly creating a transaction would provide better protection on top of the UNIQUE (namespace, sequence) constraint. Could you please add a comment above beginTx

Reply:
Done. Added a comment.